### PR TITLE
Do not request additional worker from 'yield' calls and during post-e…

### LINKED ISSUE #1704: Scheduler: Avoid unnecessary work signalling in LimitingDispatcher
To avoid liveness problems, at the end of each "limited" we double-check the queue and dispatch any new tasks from it.
At this point, a scheduler is not aware of the fact that thread it effectively free and issues one more unnecessary thread to help.

### LINKED ISSUE #1706: Scheduler: do not request additional thread for yield calls

### PR DESCRIPTION
**Pull request by qwwdfsad**
…xecution phase in LimitingDispatcher

Fixes #1704
Fixes #1706

### CODE REVIEW

**Comment by qwwdfsad** [comment]
The biggest impact here is reducing CPU usage of `ActorSelectorManager` in Ktor in idle state by a factor of two

File: kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt [hunk 1]
```diff
@@ -372,25 +372,34 @@ internal class CoroutineScheduler(
      * Dispatches execution of a runnable [block] with a hint to a scheduler whether
      * this [block] may execute blocking operations (IO, system calls, locking primitives etc.)
      *
-     * @param taskContext concurrency context of given [block]
-     * @param fair whether the task should be dispatched fairly (strict FIFO) or not (semi-FIFO)
+     * [taskContext] -- concurrency context of given [block].
+     * [tailDispatch] -- whether this [dispatch] call is the last action the (presumably) worker thread does in its current task.
```
Comment (zach-klippenstein) [inline kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt]: Just curious, why move away from using param tags?

Comment (elizarov) [inline kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt]: Please, replace `--` with `*` for a proper markdown formatting of bullet list. It does not render properly in the current shape.

File: kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt [hunk 2]
```diff
@@ -481,19 +491,19 @@ internal class CoroutineScheduler(
      * Returns `null` if task was successfully added or an instance of the
      * task that was not added or replaced (thus should be added to global queue).
      */
-    private fun submitToLocalQueue(task: Task, fair: Boolean): Task? {
-        val worker = currentWorker() ?: return task
+    private fun Worker?.submitToLocalQueue(task: Task, tailDispatch: Boolean): Task? {
+        if (this === null) return task
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt]: Style nit: `this == null`.

File: kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt (hunk 1 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt]: Lines are shorter :) 
For non-public API I personally find it more readable

Comment (elizarov) [reply kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt]: 👍 It also gives more freedom in the layout of the doc. You can have parameters documented in the course of the prose.

**Pull request merged**